File: src/lib/syncfolders.ts
/* No-sync folders — the frontend's view of which vault folders stay off sync.

   Types plus the two pure helpers the settings panel needs. Rust owns the
   decision entirely: the config file, the ownership of `.git/info/exclude`, and
   the weighing that decides whether a folder is allowed back INTO sync. These
   shapes mirror `commands::syncfolders`.

   The direction that matters: excluding a folder is free, letting one back in
   is not. Every object the sync transport carries is uploaded whole, so a
   single file past the ceiling fails the push it rides in — which is why an
   include comes back refused, with the offending files named, rather than
   quietly half-working. */

/** One folder as the panel lists it. `knownFiles`/`knownUpdated` come from the
    ghost index — the answer of whichever device actually has the folder — so a
    machine without it can still say how much is over there. */
export type SyncFolder = {
  path: string;
  excluded: boolean;
  onDisk: boolean;
  knownFiles: number;
  knownUpdated: number;
  knownCapped: boolean;
};

/** A file standing between a folder and going back into sync. */
export type OversizeFile = { path: string; size: number };

/** What weighing a folder found. Present on every include attempt, refused or
    not, so the panel can warn about a large-but-allowed one. */
export type IncludeScan = {
  files: number;
  totalBytes: number;
  oversize: OversizeFile[];
  /** Files whose size could not be read at all. These refuse the include too:
      a file the scan cannot measure is not a file it can call small enough,
      and reading "no answer" as zero bytes is how an oversize file walks past
      a size check. */
  unreadable: string[];
  limitBytes: number;
};

/** The result of a toggle. `applied: false` is a refusal, not an error —
    nothing went wrong, the folder is simply too heavy — and `scan.oversize`
    names why. */
export type SyncFolderToggle = { applied: boolean; scan: IncludeScan | null };

/** One file inside an excluded folder, as the devices that do not hold it see
    it. `path` is relative to the FOLDER, not to the vault; `mtime` is Unix
    milliseconds, `0` where the platform wouldn't say. */
export type GhostEntry = { path: string; size: number; mtime: number };

/** What one excluded folder holds, as the device that holds it last reported.
    `capped` means the listing stopped at the engine's per-folder limit and the
    count is a floor. */
export type GhostFolder = {
  updated: number;
  entries: GhostEntry[];
  capped?: boolean;
};

/** `.vault/files-index.json` — every excluded folder any device has looked at,
    keyed by vault-relative folder path. This is what lets a surface show the
    files of a folder this device does not have. */
export type GhostIndex = { version: number; folders: Record<string, GhostFolder> };

/** Bytes as a person reads them. Deliberately coarse: this labels a warning
    about whether something is worth uploading, not a file manager. */
export function howBig(bytes: number): string {
  if (bytes < 1024) return `${bytes} bytes`;
  const units = ["KB", "MB", "GB", "TB"];
  let value = bytes / 1024;
  let unit = 0;
  while (value >= 1024 && unit < units.length - 1) {
    value /= 1024;
    unit += 1;
  }
  return `${value < 10 ? value.toFixed(1) : Math.round(value)} ${units[unit]}`;
}

/** Above this an include is worth a sentence before it happens: it is allowed,
    but it is the difference between a sync that finishes and one that runs all
    evening. */
const LOUD_INCLUDE_BYTES = 1024 * 1024 * 1024;

/** What to say under a folder's name. One line, and it answers the question the
    row actually raises: excluded rows say where the files are, syncing rows say
    nothing unless there is something to warn about. */
export function folderSummary(folder: SyncFolder): string {
  if (!folder.excluded) {
    return folder.onDisk ? "Syncs to your other devices" : "Not on this device";
  }
  // A folder that is neither here nor described by any other device is one
  // nobody has put anything in yet — saying "other devices keep their copies"
  // about it would be claiming copies that do not exist.
  if (!folder.onDisk && folder.knownFiles === 0) return "Doesn't sync — nothing here yet";
  const where = folder.onDisk
    ? "Stays on this device"
    : "Not on this device — other devices keep their copies";
  if (folder.knownFiles === 0) return where;
  const count = `${folder.knownCapped ? "over " : ""}${folder.knownFiles} file${
    folder.knownFiles === 1 ? "" : "s"
  }`;
  return `${where} · ${count} known`;
}

/** The warning for an include that was allowed but is large enough to be worth
    knowing about, or null when there is nothing to say. */
export function includeWarning(scan: IncludeScan | null): string | null {
  if (!scan || scan.totalBytes < LOUD_INCLUDE_BYTES) return null;
  return `${howBig(scan.totalBytes)} across ${scan.files} file${
    scan.files === 1 ? "" : "s"
  } will now upload — the next sync will take a while.`;
}

/* The path trim itself lives in the leaf grammar module, so a pure browse model
   can normalize a folder path without importing the IPC layer. It is named here
   because this is where the folder-list vocabulary reads. */
export { normalizeFolder } from "./embedtarget.ts";

File: src-tauri/src/commands/syncfolders.rs
//! The folders this vault keeps off sync: what they are, and turning one on or
//! off.
//!
//! Three commands. Listing is a plain read; the ghost index is the same;
//! toggling is the only one that decides anything, and the one decision it makes
//! is whether a folder is allowed BACK into sync. See
//! [`crate::syncfolders::scan_for_include`] for why that direction is the
//! guarded one — files leaving sync cost nothing, files entering it are pushed
//! whole and one oversize file fails the push it rides in.

use crate::commands::history::with_history;
use crate::syncfolders::{self, IncludeScan};
use crate::{AppState, HistoryState};
use serde::Serialize;
use tauri::State;

/// One folder as the settings panel lists it.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SyncFolder {
    /// Vault-root-relative, `/`-separated.
    pub path: String,
    pub excluded: bool,
    /// Whether this device actually has the folder. A folder excluded on
    /// another device is listed here either way — that is the point of listing
    /// the config's own entries alongside what is on disk.
    pub on_disk: bool,
    /// How many files the ghost index records for it, and when. Both are the
    /// answer of whichever device last had the folder, so a device without it
    /// can still say how much is over there.
    pub known_files: usize,
    pub known_updated: u64,
    /// The listing behind `known_files` stopped at the cap.
    pub known_capped: bool,
}

/// What a toggle did — or, for an include, did not do.
///
/// A refusal is not an error: nothing went wrong, the folder is simply too
/// heavy to sync, and the panel has to name the files rather than show a
/// sentence. So the refusal comes back as a value with the scan attached, and
/// `Err` stays reserved for a vault that could not be written.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SyncFolderToggle {
    pub applied: bool,
    /// Present whenever an include was weighed, refused or not — the panel
    /// warns on a multi-gigabyte include it did allow.
    pub scan: Option<IncludeScan>,
}

#[tauri::command]
pub(crate) fn sync_folders_list(state: State<AppState>) -> Vec<SyncFolder> {
    let root = state.0.lock().unwrap().root.clone();
    let excluded = syncfolders::read_excluded(&root);
    let index = syncfolders::read_index(&root);
    let known: Vec<String> = index.folders.keys().cloned().collect();
    syncfolders::listable_folders(&root, &excluded, &known)
        .into_iter()
        .map(|path| {
            let known = index.folders.get(&path);
            SyncFolder {
                on_disk: root.join(&path).is_dir(),
                excluded: excluded.contains(&path),
                known_files: known.map(|f| f.entries.len()).unwrap_or(0),
                known_updated: known.map(|f| f.updated).unwrap_or(0),
                known_capped: known.is_some_and(|f| f.capped),
                path,
            }
        })
        .collect()
}

/// What the excluded folders hold, as the devices that have them last reported.
#[tauri::command]
pub(crate) fn sync_folders_index(state: State<AppState>) -> syncfolders::GhostIndex {
    let root = state.0.lock().unwrap().root.clone();
    syncfolders::read_index(&root)
}

/// Exclude a folder from sync, or let one back in.
///
/// Excluding takes effect at once rather than at the next idle snapshot: the
/// snapshot below is what drops the folder out of the index and records the
/// deletion other devices will read, and doing it here means the answer to "did
/// it stop syncing?" is yes by the time the switch finishes moving.
#[tauri::command]
pub(crate) fn sync_folders_set(
    state: State<AppState>,
    history: State<HistoryState>,
    folder: String,
    excluded: bool,
) -> Result<SyncFolderToggle, String> {
    let root = state.0.lock().unwrap().root.clone();
    let Some(folder) = syncfolders::normalize(&folder) else {
        return Err(format!("“{folder}” is not a folder this vault can exclude"));
    };
    // Everything under the history lock, weighing included. Outside it the scan
    // and the config write are two moments a file can grow between: an include
    // weighed as safe, a large file finishing its copy in, and the folder is
    // syncing before anybody has been told it is too heavy. Holding the lock
    // closes that window against the app's own writers. It cannot close it
    // against a hand copying files in from outside, and does not have to: the
    // scan is a gate on the user's gesture, and the transport's own per-object
    // cap remains the hard stop for whatever grows after this answer.
    with_history(&history, |hist| {
        let mut folders = syncfolders::read_excluded(&root);
        let mut scan = None;
        if excluded {
            if !folders.contains(&folder) {
                folders.push(folder.clone());
            }
        } else {
            let weighed = syncfolders::scan_for_include(&root, &folder);
            if weighed.refuses() {
                return Ok(SyncFolderToggle { applied: false, scan: Some(weighed) });
            }
            scan = Some(weighed);
            folders.retain(|f| f != &folder);
        }

        syncfolders::write_excluded(&root, &folders)?;
        // The exclusions before the snapshot: the snapshot's own untracking
        // reads the config, and everything else in the app reads `git status`,
        // so a stale exclude file here is what turns a just-excluded folder into
        // churn the next snapshot sees and cannot clear.
        hist.refresh_exclusions()?;
        let label = if excluded {
            format!("stop syncing {folder}")
        } else {
            format!("sync {folder} again")
        };
        hist.snapshot(&label)?;
        Ok(SyncFolderToggle { applied: true, scan })
    })
}
